src/dataflow/src/source/postgres.rs

```diff
@@ -157,12 +168,34 @@ impl PostgresSourceReader {
                         let a: Datum = row.get(n).into();
                         a
                     }));
-                    snapshot_tx.insert(mz_row).await?;
+                    try_recoverable!(snapshot_tx.insert(mz_row.clone()).await);
+                    try_fatal!(buffer.write(&try_fatal!(bincode::serialize(&mz_row))).await);
                 }
             }
         }
+        try_recoverable!(client.simple_query("COMMIT;").await);
+        Ok(())
+    }
 
-        client.simple_query("COMMIT;").await?;
+    /// Reverts a failed snapshot by deleting any processed rows from the dataflow.
+    async fn revert_snapshot<R: Read + Seek>(
+        &self,
+        snapshot_tx: &mut SourceTransaction<'_>,
+        mut reader: R,
+    ) -> Result<(), anyhow::Error> {
+        let rows = tokio::task::block_in_place(|| -> Result<Vec<Row>, anyhow::Error> {
+            let len = reader.seek(SeekFrom::Current(0))?;
+            reader.seek(SeekFrom::Start(0))?;
+            let mut rows = Vec::new();
+            while reader.seek(SeekFrom::Current(0))? < len {
+                rows.push(bincode::deserialize_from(&mut reader)?);
+            }
+            Ok(rows)
+        })?;
+
+        for row in rows {
```

Comment:
@petrosagg I pulled the `async` bit out of `block_in_place`, if this looks okay to you!

Reply:
Can we not do it inside the `block_in_place`? It'd be nice if the implementation never loaded the whole database in RAM even in the worst case.

Reply:
We can! I just figured out how `Handle::current().block_on()` works, fixed!